vscode/src/debugger/session.ts

```diff
@@ -0,0 +1,200 @@
+    response.body.supportsCompletionsRequest = false;
+
+    // the adapter defines two exceptions filters, one with support for conditions.
+    response.body.supportsExceptionFilterOptions = false;
+
+    // make VS Code send exceptionInfo request
+    response.body.supportsExceptionInfoRequest = false;
+
+    // make VS Code able to read and write variable memory
+    response.body.supportsReadMemoryRequest = false;
+    response.body.supportsWriteMemoryRequest = false;
+
+    response.body.supportSuspendDebuggee = false;
+    response.body.supportTerminateDebuggee = true;
+    response.body.supportsFunctionBreakpoints = true;
+
+    /* Settings that we need to eventually support: */
+
+    // make VS Code send cancel request
+    response.body.supportsCancelRequest = false;
+
+    // make VS Code use 'evaluate' when hovering over source
+    response.body.supportsEvaluateForHovers = false;
+
+    // make VS Code send the breakpointLocations request
+    response.body.supportsBreakpointLocationsRequest = false;
+    response.body.supportsDelayedStackTraceLoading = false;
+
+    // make VS Code provide "Step in Target" functionality
+    response.body.supportsStepInTargetsRequest = false;
+
+    // make VS Code send setVariable request
+    response.body.supportsSetVariable = false;
+
+    // make VS Code send setExpression request
+    response.body.supportsSetExpression = false;
+
+    // make VS Code send disassemble request
+    response.body.supportsDisassembleRequest = false;
+    response.body.supportsSteppingGranularity = false;
+
+    response.body.supportsInstructionBreakpoints = false;
+
+    this.sendResponse(response);
+
+    // since this debug adapter can accept configuration requests like 'setBreakpoint' at any time,
+    // we request them early by sending an 'initializeRequest' to the frontend.
+    // The frontend will end the configuration sequence by calling 'configurationDone' request.
+    this.sendEvent(new InitializedEvent());
+  }
+
+  /**
+   * Called at the end of the configuration sequence.
+   * Indicates that all breakpoints etc. have been sent to the DA and that the 'launch' can start.
+   */
+  protected configurationDoneRequest(
+    response: DebugProtocol.ConfigurationDoneResponse,
+    args: DebugProtocol.ConfigurationDoneArguments
+  ): void {
+    super.configurationDoneRequest(response, args);
+
+    // notify the launchRequest that configuration has finished
+    this.emit("sessionConfigurationDone");
+  }
+
+  protected async launchRequest(
+    response: DebugProtocol.LaunchResponse,
+    args: ILaunchRequestArguments
+  ) {
+    // configure DAP logging
+    logger.setup(
+      args.trace ? Logger.LogLevel.Verbose : Logger.LogLevel.Stop,
+      false
+    );
+
+    // wait until configuration has finished (configurationDoneRequest has been called)
+    const configurationDone: Promise<void> = new Promise((resolve, reject) => {
+      this.once("sessionConfigurationDone", resolve);
```

Comment:
Nice! I didn't realize the DebugSession was an EventEmitter. This is pretty clean.

Reply:
Thanks. I didn't know either, just double checked and was pleasantly surprised.